models/v3/schema/types/types.go

```diff
@@ -88,13 +91,23 @@ func (d Data) Validate() error {
 		}
 
 		rscAPIVersion := ""
+		var rscType reflect.Type
 		for i, r := range d.Resources {
 			if err := r.Validate(); err != nil {
 				return fmt.Errorf(".Resources[%d]%w", i, err)
 			}
 
+			// All ARMResource.Properties must be of the same type. This either gets the type on the
+			// first iteration or validates that the type is the same on subsequent iterations.
+			// Also makes sure that APIVersion is set on all resources if it is not set on Data and that
+			// it matches if it is set on Data.
 			if i == 0 {
 				rscAPIVersion = r.APIVersion
+				rscType = reflect.TypeOf(r.ArmResource.Properties)
```

Comment:
it seems like it should be possible to check the resource type from other fields? instead of doing a reflect here

Reply:
have another look and see if this change works.